src/sparse_blas/enum_data_types.hpp

```diff
@@ -0,0 +1,69 @@
+/***************************************************************************
+*  Copyright (C) Codeplay Software Limited
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+
+#ifndef _ONEMKL_SRC_SPARSE_BLAS_ENUM_DATA_TYPES_HPP_
+#define _ONEMKL_SRC_SPARSE_BLAS_ENUM_DATA_TYPES_HPP_
+
+#include <string>
+
+namespace oneapi::mkl::sparse::detail {
+
+enum data_type { none, int32, int64, real_fp32, real_fp64, complex_fp32, complex_fp64 };
+
+inline std::string data_type_to_str(data_type data_type) {
+    switch (data_type) {
+        case none: return "none";
+        case int32: return "int32";
+        case int64: return "int64";
+        case real_fp32: return "real_fp32";
+        case real_fp64: return "real_fp64";
+        case complex_fp32: return "complex_fp32";
+        case complex_fp64: return "complex_fp64";
+        default: return "unknown";
+    }
+}
+
+template <typename T>
+data_type get_data_type() {
+    if constexpr (std::is_same_v<T, std::int32_t>) {
+        return data_type::int32;
+    }
+    else if constexpr (std::is_same_v<T, std::int64_t>) {
+        return data_type::int64;
+    }
+    else if constexpr (std::is_same_v<T, float>) {
+        return data_type::real_fp32;
+    }
+    else if constexpr (std::is_same_v<T, double>) {
+        return data_type::real_fp64;
+    }
+    else if constexpr (std::is_same_v<T, std::complex<float>>) {
+        return data_type::complex_fp32;
+    }
+    else if constexpr (std::is_same_v<T, std::complex<double>>) {
+        return data_type::complex_fp64;
+    }
+    else {
+        static_assert(false, "Unsupported type");
```

Comment:
If `none` is not a "supported type" or if `none` is not expected to behave differently from `default` case, why do we have it around?

Reply:
The `none` value is used in 2 cases:

- When initializing the generic_container with a [default constructor](https://github.com/oneapi-src/oneMKL/pull/500/files#diff-94c98d6e052fa1c50afe95a3973550ec75c0a51fcdea2d80b5de68886c9f83f6R51). The `none` value means that a type hasn't been set yet. This is useful to detect an internal error if `get_data_type` was used with an untyped container.
- In [`generic_dense_handle::get_int_type`](https://github.com/oneapi-src/oneMKL/pull/500/files#diff-94c98d6e052fa1c50afe95a3973550ec75c0a51fcdea2d80b5de68886c9f83f6R119). A dense handle does not have an int type but the method is useful to check that a set of generic handles have compatible type (see [check_all_containers_compatible](https://github.com/oneapi-src/oneMKL/pull/500/files#diff-94c98d6e052fa1c50afe95a3973550ec75c0a51fcdea2d80b5de68886c9f83f6R299)). In this case a `none` type is considered compatible with any other integer type.

Reply:
Resoving old comment